diffkemp/simpll/passes/FieldAccessFunctionGenerator.cpp

```diff
@@ -0,0 +1,175 @@
+                        DIL = Inst.getDebugLoc().get();
+                        InstStack.push_back(&Inst);
+                    }
+                } else if ((isa<GetElementPtrInst>(Inst) ||
+                           isa<CastInst>(Inst)) && !InstStack.empty()) {
+                   if (Inst.getDebugLoc().get() != DIL) {
+                       // Stop building stack.
+                       processStack(InstStack, Mod);
+                       InstStack.clear();
+                   } else {
+                       // Add instruction to stack.
+                       InstStack.push_back(&Inst);
+                   }
+                } else {
+                    // Wrong instruction type. Stop building stack.
+                    processStack(InstStack, Mod);
+                    InstStack.clear();
+                }
+            }
+            // Avoid processing instruction groups that contain instructions in
+            // two or more basic blocks.
+            processStack(InstStack, Mod);
+            InstStack.clear();
+        }
+    }
+
+    return PreservedAnalyses();
+}
+
+/// First checks whether the instructions in the stack can be moved away from
+/// the original function without causing an error, then creates the abstraction
+/// function, moves the instructions into it, modify the input and output of the
+/// block to match the argument and returns value of the function and inserts
+/// the call to the newly create abstraction instead of the instruction block to
+/// the original function.
+void FieldAccessFunctionGenerator::processStack(
+        const std::vector<Instruction *> &Stack, Module &Mod) {
+    if (Stack.empty())
+        // Do not attempt to process empty stacks, that would cause an error.
+        return;
+
+    // Check if all the instruction (besides the last one, of course) are used
+    // only by other instructions in the stack.
+    // If not, do not generate an abstraction, since this would break the code.
+    for (Instruction *Inst : Stack) {
+        if (Inst == Stack.back())
+            continue;
+        for (User *U : Inst->users()) {
+            if (std::find(Stack.begin(), Stack.end(), U) == std::end(Stack))
+                return;
+        }
+    }
+
+    // Check if all operands in the code are values included in the set created
+    // in the previous step. If not, it would be technically possible to do this
+    // transformation, but this is not supported by this implementation.
+    for (Instruction *Inst : Stack) {
+        if (Inst == Stack.front()) {
+            // The GEP is guaranteed to have all indices constant and the first
+            // value is extracted.
+            continue;
+        }
+        for (Value *Op : Inst->operands()) {
+            if (isa<ConstantData>(Op))
+                // Constants are fine, they can be moved without problem.
+                continue;
+            if (std::find(Stack.begin(), Stack.end(), Op) != std::end(Stack))
```

Comment:
Use the same style as in the above loop:

    if (std::find(...) == std::end(...))
        return;

It is confusing when you use different styles for the same functionality.